cassiopeia/plotting/itol_utilities.py

```diff
@@ -0,0 +1,619 @@
+            temporary_directory,
+            indel_colors,
+            indel_priors,
+            random_state,
+        )
+
+    for meta_item in meta_data:
+        if meta_item not in cassiopeia_tree.cell_meta.columns:
+            raise iTOLError("Meta data item not in CassiopeiaTree cell meta.")
+
+        values = cassiopeia_tree.cell_meta[meta_item]
+
+        if pd.api.types.is_numeric_dtype(values):
+            files.append(create_gradient_from_df(
+                values, cassiopeia_tree, f"{tree_name}.{meta_item}", temporary_directory,
+            ))
+
+        if pd.api.types.is_string_dtype(values):
+            colors = palette[: len(values.unique())]
+            colors = [hex_to_rgb(color) for color in colors]
+            colormap = dict(zip(np.unique(values), colors))
+
+            files.append(create_colorbar(
+                values,
+                cassiopeia_tree,
+                colormap,
+                f"{tree_name}.{meta_item}",
+                temporary_directory,
+                create_legend=include_legend,
+            ))
+
+    for _file in files:
+        itol_uploader.add_file(_file)
+
+    itol_uploader.params["treeName"] = tree_name
+    itol_uploader.params["APIkey"] = api_key
+    itol_uploader.params["projectName"] = project_name
+
+    good_upload = itol_uploader.upload()
+    if not good_upload:
+        raise iTOLError(itol_uploader.comm.upload_output)
+
+    print("iTOL output: " + str(itol_uploader.comm.upload_output))
+    print("Tree Web Page URL: " + itol_uploader.get_webpage())
+    print("Warnings: " + str(itol_uploader.comm.warnings))
+
+    tree_id = itol_uploader.comm.tree_id
+
+    itol_exporter = ItolExport()
+
+    # set parameters:
+    itol_exporter.set_export_param_value("tree", tree_id)
+    itol_exporter.set_export_param_value("format", file_format)
+    if rect:
+        # rectangular tree settings
+        itol_exporter.set_export_param_value("display_mode", 1)
+    else:
+        # circular tree settings
+        itol_exporter.set_export_param_value("display_mode", 2)
+        itol_exporter.set_export_param_value("arc", 359)
+        itol_exporter.set_export_param_value("rotation", 270)
+
+    itol_exporter.set_export_param_value("leaf_sorting", 1)
+    itol_exporter.set_export_param_value("label_display", 0)
+    itol_exporter.set_export_param_value("internal_marks", 0)
+    itol_exporter.set_export_param_value("ignore_branch_length", 1)
+
+    itol_exporter.set_export_param_value(
+        "datasets_visible", ",".join([str(i) for i in range(len(files))])
+    )
+
+    itol_exporter.set_export_param_value("horizontal_scale_factor", 1)
+
+    # export!
+    itol_exporter.export(export_filepath)
+
+    # remove intermediate files
+    shutil.rmtree(temporary_directory)
+
+
+def create_gradient_from_df(
+    df: pd.Series,
+    tree: CassiopeiaTree,
+    dataset_name: str,
+    output_directory: str = "./tmp/",
+    color_min: str = "#ffffff",
+    color_max: str = "#000000",
```

Comment:
These fields are inaccesible from the outer function, should we add them to upload_and_export_itol

Reply:
Additionally, should we allow the user to specify min and max colors for the random color/indel generation?

Reply:
I don't think we should have th user specify min/max colors for the random color/indel generation -- it gets too hectic there. Plus, we already thought a lot about the random color generation a while ago. I think it works well.

I'd like to ideally pass in color bounds for the numerical data, but I don't see an elegant way of doing this. I think I should just add this as a todo.

Reply:
That sounds like a plan